src/librustdoc/passes/html_tags.rs

```diff
@@ -0,0 +1,144 @@
+use super::{span_of_attrs, Pass};
+use crate::clean::*;
+use crate::core::DocContext;
+use crate::fold::DocFolder;
+use crate::html::markdown::opts;
+use core::ops::Range;
+use pulldown_cmark::{Event, Parser};
+use rustc_feature::UnstableFeatures;
+use rustc_session::lint;
+
+pub const CHECK_INVALID_HTML_TAGS: Pass = Pass {
+    name: "check-invalid-html-tags",
+    run: check_invalid_html_tags,
+    description: "detects invalid HTML tags in doc comments",
+};
+
+struct InvalidHtmlTagsLinter<'a, 'tcx> {
+    cx: &'a DocContext<'tcx>,
+}
+
+impl<'a, 'tcx> InvalidHtmlTagsLinter<'a, 'tcx> {
+    fn new(cx: &'a DocContext<'tcx>) -> Self {
+        InvalidHtmlTagsLinter { cx }
+    }
+}
+
+pub fn check_invalid_html_tags(krate: Crate, cx: &DocContext<'_>) -> Crate {
+    if !UnstableFeatures::from_environment().is_nightly_build() {
+        krate
+    } else {
+        let mut coll = InvalidHtmlTagsLinter::new(cx);
+
+        coll.fold_crate(krate)
+    }
+}
+
+const ALLOWED_UNCLOSED: &[&str] = &[
+    "area", "base", "br", "col", "embed", "hr", "img", "input", "keygen", "link", "meta", "param",
+    "source", "track", "wbr",
+];
+
+fn drop_tag(
+    tags: &mut Vec<(String, Range<usize>)>,
+    tag_name: String,
+    range: Range<usize>,
+    f: &impl Fn(&str, &Range<usize>),
+) {
+    if let Some(pos) = tags.iter().position(|(t, _)| *t == tag_name) {
```

Comment:
I thought of another thing: what happens if you have nested tags? I think this will mistakenly warn on the following code:

```rust
/// <div>
///   <p> //~ WARNING <p> not closed
///      <div></div> //~ WARNING <div> not closed
///   </p> //~ WARNING <p> has no opening tag
/// </div>
```
Try using a reverse iterator instead so it uses the most recently opened tag:
```suggestion
    if let Some(pos) = tags.iter().rev().position(|(t, _)| *t == tag_name) {
```
Can you also add my code snippet as a test?

Reply:
Great catch! I completely forgot the `rev` part...